staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/stdlib.go

```diff
@@ -0,0 +1,616 @@
+				if !ok {
+					// Should never happen for type checked CEL programs
+					return types.NoSuchOverloadErr()
+				}
+				acc = sum
+			} else {
+				acc = nextAdder
+			}
+		}
+		return acc.(ref.Val)
+	}
+}
+
+func min() functions.UnaryOp {
+	return cmp("min", types.IntOne)
+}
+
+func max() functions.UnaryOp {
+	return cmp("max", types.IntNegOne)
+}
+
+func cmp(opName string, opPreferCmpResult ref.Val) functions.UnaryOp {
+	return func(val ref.Val) ref.Val {
+		var result traits.Comparer
+		iterable, ok := val.(traits.Iterable)
+		if !ok {
+			return types.NoSuchOverloadErr()
+		}
+		for it := iterable.Iterator(); it.HasNext() == types.True; {
+			next := it.Next()
+			nextCmp, ok := next.(traits.Comparer)
+			if !ok {
+				// Should never happen for type checked CEL programs
+				return types.NoSuchOverloadErr()
+			}
+			if result == nil {
+				result = nextCmp
+			} else {
+				cmp := result.Compare(next)
+				if cmp == opPreferCmpResult {
+					result = nextCmp
+				}
+			}
+		}
+		if result == nil {
+			return types.NewErr("%s called on empty list", opName)
+		}
+		return result.(ref.Val)
+	}
+}
+
+func indexOf(list ref.Val, item ref.Val) ref.Val {
+	lister, ok := list.(traits.Lister)
+	if !ok {
+		return types.NoSuchOverloadErr()
+	}
+	sz := lister.Size().Value().(int64)
+	for i := int64(0); i < sz; i++ {
+		if lister.Get(types.Int(i)).Equal(item) == types.True {
+			return types.Int(i)
+		}
+	}
+	return types.Int(-1)
+}
+
+func lastIndexOf(list ref.Val, item ref.Val) ref.Val {
+	lister, ok := list.(traits.Lister)
+	if !ok {
+		return types.NoSuchOverloadErr()
+	}
+	sz := lister.Size().Value().(int64)
+	for i := sz - 1; i >= 0; i-- {
+		if lister.Get(types.Int(i)).Equal(item) == types.True {
+			return types.Int(i)
+		}
+	}
+	return types.Int(-1)
+}
+
+func find(strVal ref.Val, regexVal ref.Val) ref.Val {
+	str, ok := strVal.Value().(string)
+	if !ok {
+		return types.NoSuchOverloadErr()
+	}
+	regex, ok := regexVal.Value().(string)
+	if !ok {
+		return types.NoSuchOverloadErr()
+	}
+	re, err := regexp.Compile(regex)
+	if err != nil {
+		return types.NewErr("Illegal regex: %v", err.Error())
+	}
+	result := re.FindString(str)
+	return types.String(result)
+}
+
+func findAll(args ...ref.Val) ref.Val {
+	argn := len(args)
+	if argn < 2 || argn > 3 {
+		return types.NoSuchOverloadErr()
+	}
+	str, ok := args[0].Value().(string)
+	if !ok {
+		return types.NoSuchOverloadErr()
+	}
+	regex, ok := args[1].Value().(string)
+	if !ok {
+		return types.NoSuchOverloadErr()
+	}
+	n := int64(-1)
+	if argn == 3 {
+		n, ok = args[2].Value().(int64)
+		if !ok {
+			return types.NoSuchOverloadErr()
+		}
+	}
+
+	re, err := regexp.Compile(regex)
+	if err != nil {
+		return types.NewErr("Illegal regex: %v", err.Error())
+	}
+
+	result := re.FindAllString(str, int(n))
+
+	return types.NewStringList(types.DefaultTypeAdapter, result)
+}
+
+func stringToUrl(arg ref.Val) ref.Val {
+	s := arg.Value().(string)
+	u, err := url.Parse(s)
```

Comment:
url.Parse is surprisingly lenient... https://go.dev/play/p/hDhM82vpDyA

was this expected to only allow urls that specify scheme and host? should it allow empty ports? `file:///` urls?

Reply:
Very true.

We could use url.ParseRequestURI instead. ParseRequestURI is what we use for the OpenAPIv3 "uri" format and is slightly more strict (although still pretty lenient IMHO)

The catch is that ParseRequestURI doesn't support fragments right. It allows them, but it doesn't handle them right. So if they're present the end up appended to either the path or the query, depending on which they are adjacent to.

One option would be to guard the conversion with ParseRequestURI but then parse using url.Parse.


This brings up something I had not considered when writing the KEP section-- I don't think there is any hope of writing this URL library such that it could be reimplemented in other languages. This means that we're making our CEL "flavor" a lot less portable by introducing this to the library.

With that in mind I'm currently in favor of dropping the URL support. Thoughts?

Reply:
I think anything we do should be coherent with our `format: url` support, so at least guarding with `ParseRequestURI` seems like an improvement

I didn't have super concrete use cases in mind for the url bits. Extracting scheme/host/path/query seem like they should be reasonable to do cross-platform. User auth info and fragment seem significantly less useful to me.

Reply:
if we also dropped the url → string function, then the only purpose of converting a string to a url would be to extract scheme/host/port/path, which seems more likely to be consistent cross-platform than url generation

Reply:
IIRC, @deads2k was more in favor of providing URL functions? maybe he can comment on the use cases and potentially narrowed scope

Reply:
I've added the `ParseRequestURI` guard and reduced to scheme/host/hostname/port/path/query and isUrl and string->url conversion.  This might be manageable. If there are any ideas on how we can narrow it any further I'm happy to keep chopping.